virtio-queue: Error out early on invalid available ring index

likebreath opened the pull request:
The number of descriptor chain heads to process should always be smaller or equal to the queue size, as the driver should never ask the VMM to process a available ring entry more than once. Checking and reporting such incorrect driver behavior can avoid potential hanging and Denial-of-Service from VMMs.

Signed-off-by: Bo Chen <chen.bo@intel.com>

### Summary of the PR

*Please summarize here why the changes in this PR are needed.*

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR are signed (with `git commit -s`), and the commit
  message has max 60 characters for the summary and max 75 characters for each
  description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] Any newly added `unsafe` code is properly documented.

likebreath commented:
/cc @rbradford @sboeuf

andreeaflorescu commented on crates/virtio-queue/src/queue.rs:
```diff
@@ -611,8 +611,18 @@ impl QueueOwnedT for Queue {
+        // be smaller or equal to the queue size, as the driver should
+        // never ask the VMM to process a available ring entry more than
+        // once. Checking and reporting such incorrect driver behavior
+        // can avoid potential hanging and Denial-of-Service from VMMs.
```
Do you have a reproduction for this DoS? When does this happen?

When you call `next` on the returned object from this function we already check that it is less then size because we're doing a modulo: https://github.com/rust-vmm/vm-virtio/blob/7e203db6ed044217acd80dec5967c15d979491f8/crates/virtio-queue/src/queue.rs#L750.

According to the spec the driver is allowed to pass a number that is larger than QUEUE_SIZE: https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-380006
"idx field indicates where the driver would put the next descriptor entry in the ring (modulo the queue size)."

rbradford replied:
Not sure the logic in the `if` block is correct (because of wrapping) but the goal is to avoid iterating and processing over the same descriptors over and over again. 

e.g. if the queue size was `n` and guest set the next descriptor  to `current + 1000 * n` it would force the virtio device to consider each descriptor in the array 1000 times. This would effectively DoS the thread processing the queue especially if the computation on the descriptor is costly. This was found with fuzzing.

rbradford replied:
This manifests itself by `.next()` continuing to return descriptor chains until `next_avail` and `last_index` are equal: https://github.com/rust-vmm/vm-virtio/blob/main/crates/virtio-queue/src/queue.rs#L743

andreeaflorescu replied:
@rbradford @likebreath do you think you can work on a unit test that exemplifies this scenario? We can use it as a regression test for the fix. I am trying to do that myself, but I am still not very sure how this can happen. The guest can flood the VMM with requests nevertheless, so some sort of a rate limiter needs to be implemented at the VMM side nevertheless. If we can improve the situation in any case at the device level, that would be great as well.

rbradford replied:
> The guest can flood the VMM with requests nevertheless, so some sort of a rate limiter needs to be implemented at the VMM side nevertheless. If we can improve the situation in any case at the device level, that would be great as well.

A well behaved guest would never do this because they wouldn't want the same descriptors handled multiple times. I don't know if we could also use the used ring to identify this troublesome behaviour.

rbradford replied:
> I don't know if we could also use the used ring to identify this troublesome behaviour.

I'm pretty sure we can't because it's only written by the device.

andreeaflorescu replied:
Hmm, I think the fix here might be to do the mod when we create the `AvailIter` because then the if condition in `next` will correctly return None.

dianpopa replied:
> Do you have a reproduction for this DoS? When does this happen?
> 
> When you call `next` on the returned object from this function we already check that it is less then size because we're doing a modulo:
> 
> https://github.com/rust-vmm/vm-virtio/blob/7e203db6ed044217acd80dec5967c15d979491f8/crates/virtio-queue/src/queue.rs#L750
> 
> .
> According to the spec the driver is allowed to pass a number that is larger than QUEUE_SIZE: https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-380006 "idx field indicates where the driver would put the next descriptor entry in the ring (modulo the queue size)."

Indeed the idx field (which is read from guest memory  through the `self.avail_idx()` fn) can be larger than QUEUE_SIZE but my understanding is that the current fix puts a limit on the number of available descriptors (i.e idx - self.next_avail). 
If we allow the number of available descriptors to be higher than queue size, we end up processing some descriptors more than once.

rbradford replied:
> If we allow the number of available descriptors to be higher than queue size, we end up processing some descriptors more than once.

Exactly. I can't think of any reason why a well behaved guest would want that.

andreeaflorescu replied:
> Hmm, I think the fix here might be to do the mod when we create the `AvailIter` because then the if condition in `next` will correctly return None.

This is not actually correct because the spec says that the [idx grows and wraps around at u16::MAX](https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-55003r55). I've mis-read the spec. The modulo was referring to writing elements in the vring.

I also looking how other projects are addressing this problem. We can't directly do `idx - self.next_avail` because that operation might underflow and cause a panic. So we must do a checked_sub if we want to check that the same descriptor is not processed multiple times.

rbradford replied:
> We can't directly do `idx - self.next_avail` because that operation might underflow and cause a panic. 

Both these are `Wrapping<u16>` so I think that is it safe.

andreeaflorescu replied:
Looks like QEMU has the same check in place: https://github.com/qemu/qemu/blob/master/hw/virtio/virtio.c#L966

rbradford replied:
> Looks like QEMU has the same check in place: https://github.com/qemu/qemu/blob/master/hw/virtio/virtio.c#L966

Cool. Do you think this is the right place or should it be in the implementation of `.next()` ?

andreeaflorescu replied:
> > We can't directly do `idx - self.next_avail` because that operation might underflow and cause a panic.
> 
> Both these are `Wrapping<u16>` so I think that is it safe.

Right, that is safe because it will return u16::MAX which is going to be > queue size irrespective of the queue size set by the driver.

andreeaflorescu replied:
> Cool. Do you think this is the right place or should it be in the implementation of `.next()` ?

I am checking to see if there is any other code path that could trigger this behavior outside of `iter`. It might make sense to have it in the constructor of `AvailIter`.

rbradford replied:
So @likebreath looks like the actions that you need to do on this PR are address the comment I made about the commit message/comment and add a unit tests to ensure that invalid and only invalid ranges generate an error.

rbradford commented on crates/virtio-queue/src/queue.rs:
```diff
@@ -611,8 +611,18 @@ impl QueueOwnedT for Queue {
-            .map(move |idx| AvailIter::new(mem, idx, self))
+        let idx = self.avail_idx(mem.deref(), Ordering::Acquire)?;
+
+        // The number of descriptor chain heads to process should always
```
I don't think this comment or that in the commit message is quite right. The VMM isn't the part that can cause the DoS it's the guest.

likebreath replied:
Thank you for letting me know the message was misleading. That was not what I intended to say. 

Both the comments and commit message are updated. PTAL.

andreeaflorescu commented on crates/virtio-queue/src/queue.rs:
```diff
@@ -1238,18 +1248,8 @@ mod tests {
-                break;
-            }
-        }
+        // Invalid available ring index
```
We should add some additional tests here:
1. A regression that fails before the fix and passes afterwards. We can write a unit test for the scenario described by Rob where the 
2. A test to check that you can still handle exactly queue size descriptor chains.

likebreath replied:
A new unit test is added from a separate commit. Anything I am missing, please let me know. Thank you!

likebreath commented:
Thank you all the the feedbacks. All comments should be addressed. PTAL. 

A summary of changes:
1. Moved the check for invalid available ring index to the constructor of `AvailIter`;
2. Clarified the commit message and comments;
3. Added a related unit test;
